### PR TITLE
WIP: Wayland z layering

### PR DESCRIPTION
**Pull request by jwijenbergh**
Some TODOs here and there, just want this PRed now so people can test.

Also tests still TODO of course

### CODE REVIEW

File: libqtile/backend/wayland/core.py [hunk 1]
```diff
@@ -308,11 +353,25 @@ def __init__(self) -> None:
         self.layer_trees = [
             SceneTree.create(self.windows_tree),  # Background
             SceneTree.create(self.windows_tree),  # Bottom
+            SceneTree.create(self.windows_tree),  # Keep below windows
             SceneTree.create(self.windows_tree),  # Regular windows
+            SceneTree.create(
+                self.windows_tree
+            ),  # Keep above windows (floats go here if floats_kept_above
```
Comment (jwijenbergh) [inline libqtile/backend/wayland/core.py:360]: holy crap black is ugly

File: libqtile/backend/wayland/window.py [hunk 2]
```diff
@@ -952,7 +1021,7 @@ def place(
         respect_hints: bool = False,
     ) -> None:
         if above:
-            self.bring_to_front()
+            self.move_up()
```
Comment (jwijenbergh) [inline libqtile/backend/wayland/window.py:1024]: @elParaguayo is this correct?

File: libqtile/backend/wayland/window.py [hunk 3]
```diff
@@ -686,10 +698,67 @@ def enable_fullscreen(self) -> None:
...
+        self.reparent(self.core.keep_below_window_tree)
+
+    @expose_command()
+    def keep_above(self) -> None:
+        self.reparent(self.core.keep_above_window_tree)
+
+    @expose_command()
+    def move_up(self, force: bool = False) -> None:
+        if force and self.tree_layer is self.core.keep_below_window_tree:
+            self.reparent(self.get_new_layer(self._float_state))
```
Comment (jwijenbergh) [inline libqtile/backend/wayland/window.py:719]: @elParaguayo does move_up or down with force also remove it from bring_to_front layer?

File: libqtile/backend/wayland/xwindow.py [hunk 4]
```diff
@@ -478,6 +475,7 @@ def unhide(self) -> None:
                 self.tree.node.set_position(self.borderwidth, self.borderwidth)
 
             self.container.node.set_enabled(enabled=True)
+            # TODO: is this correct
             self.bring_to_front()
```
Comment (jwijenbergh) [inline libqtile/backend/wayland/xwindow.py:479]: I have to check all bring to front calls in the wayland code base

File: libqtile/backend/wayland/window.py (hunk 2 above)
Comment (jwijenbergh) [reply libqtile/backend/wayland/window.py:1024]: I am confused what needs to happen when above is set in `place`

File: libqtile/backend/wayland/window.py (hunk 3 above)
Comment (elParaguayo) [reply libqtile/backend/wayland/window.py:719]: x11 doesn't have a `bring_to_front` layer. In x11, `bring_to_front` ignores layers and brings the window to the front. Once it loses focus, it should be restacked according to its layer.

The behaviour of `force` is as follows. If this is `True`, `move_up` will move the window out of the `kept_below` layer group, and `move_down` will move the window out of the `kept_above` layer group. If `False`, the windows can only move within their respective layer groups.

File: libqtile/backend/wayland/window.py [hunk 5]
```diff
@@ -686,10 +698,71 @@ def enable_fullscreen(self) -> None:
...
+        self.container.node.reparent(layer)
+
     @expose_command()
-    def bring_to_front(self) -> None:
+    def keep_below(self) -> None:
+        self.reparent(self.core.keep_below_window_tree)
+
+    @expose_command()
+    def keep_above(self) -> None:
+        self.reparent(self.core.keep_above_window_tree)
```
Comment (elParaguayo) [inline libqtile/backend/wayland/window.py:714]: In X11, we have a `kept_above/below` property. We should add that here. 

In addition, the X11 method signature includes an `enable` argument. If this is set, then `kept_above/below` is set to that value. If it's `None` then the state is toggled.

File: libqtile/backend/wayland/window.py (hunk 2 above)
Comment (elParaguayo) [reply libqtile/backend/wayland/window.py:1028]: I think `move_up` is ok here - but let's see what happens in the tests ;)

**Review by elParaguayo — COMMENTED** [review]
Have started having a go with this. Some thoughts below.

Also noticed that rofi opened behind floating windows. I then had a coredump...

**Pull request closed without merging**